### DISCUSSION TITLE
IDictionary - ConcurrentDictionary TryAdd throws Exception

### DISCUSSION THREAD

**Discussion by Flatscha** [dotnet/runtime#88471] [General]
### Description

If 2 processes call TryAdd on the same dictionary TryAdd will throw the Exception on the second one
```
Exception Info: System.ArgumentException: The key already existed in the dictionary.
   at System.Collections.Concurrent.ConcurrentDictionary`2.System.Collections.Generic.IDictionary<TKey,TValue>.Add(TKey key, TValue value)
   at System.Collections.Generic.CollectionExtensions.TryAdd[TKey,TValue](IDictionary`2 dictionary, TKey key, TValue value)
```

### Reproduction Steps

Create a dictionary 
        `IDictionary<string, object> Dict = new ConcurrentDictionary<string, object>();`
Start two tasks that call TryAdd onto that dictionary at the same time

### Expected behavior

TryAdd should not throw any Exceptions besides `ArgumentNullException`

### Actual behavior

TryAdd throws `System.ArgumentException: The key already existed in the dictionary.`

### Configuration

- Which version of .NET is the code running on? 
   **.NET 7.0**
- What OS and version, and what distro if applicable?
   **Windows 10 Enterprise / Windows 10 IoT / Windows Server 2016 Standard**
- What is the architecture (x64, x86, ARM, ARM64)?
   **x64**
- Do you know whether it is specific to that configuration?
  **I don't think so**
- If you're using Blazor, which web browser(s) do you see this issue in?
   **newest release of chrome**

### Other information

There should be a catch which returns false in the TryAdd method

**Comment by ghost** [comment]
Tagging subscribers to this area: @dotnet/area-system-collections
See info in [area-owners.md](https://github.com/dotnet/runtime/blob/main/docs/area-owners.md) if you want to be subscribed.
<details>
<summary>Issue Details</summary>
<hr />

### Description

If 2 processes call TryAdd on the same dictionary TryAdd will throw the Exception on the second one
```
Exception Info: System.ArgumentException: The key already existed in the dictionary.
   at System.Collections.Concurrent.ConcurrentDictionary`2.System.Collections.Generic.IDictionary<TKey,TValue>.Add(TKey key, TValue value)
   at System.Collections.Generic.CollectionExtensions.TryAdd[TKey,TValue](IDictionary`2 dictionary, TKey key, TValue value)
```

### Reproduction Steps

Create a dictionary 
        `IDictionary<string, object> Dict = new ConcurrentDictionary<string, object>();`
Start two tasks that call TryAdd onto that dictionary at the same time

### Expected behavior

TryAdd should not throw any Exceptions besides `ArgumentNullException`

### Actual behavior

TryAdd throws `System.ArgumentException: The key already existed in the dictionary.`

### Regression?

_No response_

### Known Workarounds

_No response_

### Configuration

- Which version of .NET is the code running on? 
   **.NET 7.0**
- What OS and version, and what distro if applicable?
   **Windows 10 Enterprise / Windows 10 IoT / Windows Server 2016 Standard**
- What is the architecture (x64, x86, ARM, ARM64)?
   **x64**
- Do you know whether it is specific to that configuration?
  **I don't think so**
- If you're using Blazor, which web browser(s) do you see this issue in?
   **newest release of chrome**

### Other information

There should be a catch which returns false in the TryAdd method

<table>
  <tr>
    <th align="left">Author:</th>
    <td>Flatscha</td>
  </tr>
  <tr>
    <th align="left">Assignees:</th>
    <td>-</td>
  </tr>
  <tr>
    <th align="left">Labels:</th>
    <td>

`area-System.Collections`, `untriaged`

</td>
  </tr>
  <tr>
    <th align="left">Milestone:</th>
    <td>-</td>
  </tr>
</table>
</details>

**Comment by Tornhoof** [comment]
That's because ``TryAdd`` will be resolved as the extension method on IDictionary and that one basically calls ``ContainsKey`` and then ``Add``.
To use the ConcurrentDictionary one, don't cast your dictionary to IDictionary.

**Reply by DaZombieKiller** [reply]
This makes me wonder if `CollectionExtensions.TryAdd` could special-case `ConcurrentDictionary` here, similar to how LINQ special-cases `List<T>` and others for perf reasons.

**Reply by LeaFrock** [reply]
This inconsistent behavior between extension methods and self-owned methods is indeed confusing. A special-case is also not very satisfying... What I'm curious is why [IDictionary.Add of ConcurrentDictionary](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs#L1523) need to throw an exception if `TryAdd` is false?

**Reply by stephentoub** [reply]
> What I'm curious is why [IDictionary.Add of ConcurrentDictionary](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs#L1523) need to throw an exception if TryAdd is false?

Because that's the documented behavior of `IDictionary<TKey, TValue>.Add` and how it communicates that the key already exists and that the supplied value is being dropped on the floor.

**Reply by LeaFrock** [reply]
Oh, thanks. I see the [doc](https://learn.microsoft.com/en-us/dotnet/api/system.collections.idictionary.add).

Therefore, this problem is just by design, which increases a little bit learning cost for developers 😿.

This drives me turning to support a special case for `ConcurrentDictionary` as [CollectionExtensions.TryAdd](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.collectionextensions.tryadd) doesn't mention it will throw an exception, and it's also a convention with `Try-*`.

**Reply by huoyaoyuan** [reply]
Also, a layering or implementation-wise issue: `CollectionExtensions` lives in the same assembly of base `IDictionary` interface definition (the CoreLib). `ConcurrentDictionary` is in the concurrent collection assembly, which is built on top of the base assembly.

This is a case where default interface methods can perform better than extension methods.

**Comment by rmannibucau** [comment]
up, any way default impls are supported in default extensions? this is really bothering ina big codebase where the actual impl type cannot be propagated for good reasons

**Reply by huoyaoyuan** [reply]
It can be error prone to use concurrent collections with any method designed for non-concurrent collections. A good solution should be a roslyn analyzer that catches any attempt to cast concurrent collections to non-concurrent interface.

**Reply by rmannibucau** [reply]
@huoyaoyuan an analyzer is *not* a solution, a concurrent dictionary is a idictionary so until it is invalid to use it as such - and therefore it should be deprecated and "duplicated" to not implement idictionary in terms of API - then it is perfectly valid to use lighter constraints with implementations guarantees - this is all the object model and inheritance/interfaces are about even if dotnet has some constraints on that. Concurrent dictionary is often used as a cache and if you integrate with a 3rd party which uses a IDictionary you are then enforced to lock which defeats the goal but there is no reason to push back to libs to support concurrent dictionary which has other side effects in terms of API requirements. So overall the support of a proper polymorphic dictionary is pretty easy and trivial and the issue only comes from a partial extension implementation which is easy to fix so better to just do it IMHO.

**Reply by huoyaoyuan** [reply]
I'm not aware of anything being easy or trivial. Extension methods are not a part of the type and are fully resolved by overload resolution at compile time. All the polymorphic stuffs are unrelated to extension methods.

To achieve a better story for this, *extension interface implementation* is required, which allows you "attach" interface implementation to existing types and thus being polymorphic. However, this feature is still in long term goal and no design conclusion has been done.

**Reply by rmannibucau** [reply]
well this is trivial since built-in and in the "same bundle" so the minimum is to cover what is produced by its own code - it is what I meant.
There already are some internals hacks like that in the runtime so this one is really not hurting and solving a common issue dev can't fix by design.

Agree a better extension polymorphism solution would be great but for this particular issue the fix is easy and a huge quick win IMHO.

**Reply by huoyaoyuan** [reply]
> There already are some internals hacks like that in the runtime so this one is really not hurting

No, the runtime only use internal hacks in very limited scenarios, mostly within the same layer and components. The API surface of `CollectionExtensions` and `ConcurrentDictionary` are in different layers. Hacking may solve only this one method but not any others, especially LINQ usages.

**Reply by rmannibucau** [reply]
I get that (even if the layers are close enough for me to be worth a hack removed when a good design for that issue is found) but what's the solution then? I could go with the analyzer solution if it traverses libraries for such an usage (it doesn't by desfault and is limited to your own code for most impls AFAIK) but it sounds like it will slow down a lot the build for something which is really intended to work OOTB - main issue being fro within an IDE you rarely notice a method comes from an extension or not.
So still think a short term solution should happen for known overriden methods of the runtime namespaces for a better interoperability/consistency between them.

**Reply by colejohnson66** [reply]
@huoyaoyuan I take it that it's not possible for `CollectionExtensions.TryAdd` to special case `dict is ConcurrentDictionary<TKey,TValue>`?

**Reply by huoyaoyuan** [reply]
> I take it that it's not possible for `CollectionExtensions.TryAdd` to special case `dict is ConcurrentDictionary<TKey,TValue>`?

It is possible in *current implementation shape* because both types are folded into CoreLib.

But it's not possible for *current API contract*, because `CollectionExtensions` is exposed in `System.Collections` while `ConcurrentDictionary<TKey,TValue>` is in `System.Collections.Concurrent`, and the former shouldn't have knowledge about the later.

**Reply by stephentoub** [reply]
> It is possible in current implementation shape because both types are folded into CoreLib.

The implementation of ConcurrentDictionary is in System.Collections.Concurrent.

**Reply by huoyaoyuan** [reply]
Ah, I got confused with `ConcurrentQueue`. So it's not possible implementation-wise.

**Reply by rmannibucau** [reply]
Even if they are in different assemblies it should be possible, worse case with an internal (copy on write pattern to avoid locks) Dictionary<Predicate<object>, TryAddDelegate> the concurrent dictionary can register statically (when class is loaded) in the extension, this way it will behave as expected even if modules are not the same and this dictionary can be opened for concurrent assembly but not end users to enable to break it when a proper solution is found.
So clearly there are multiple solution without even reflection and with low enough cost - dictionary option can even just be a dedicated static attribute if concurrent dictionary is the only case, overhead will be pretty much 0 and it will be functional for end users.

Nothing is never impossible in dev ;)

**Reply by huoyaoyuan** [reply]
No, such hack is totally unacceptable and inactionable. There are strictly NO internal access between assemblies.

> So clearly there are multiple solution without even reflection and with low enough cost

There are 0 acceptable solutions. The overhead is unacceptable too because it breaks inlining etc.

> Nothing is never impossible in dev ;)

"Rebuild the world" is always theoretically possible, but won't ever be considered in real world engineering.

**Reply by rmannibucau** [reply]
@huoyaoyuan at some point the question is what is the worse: breaking code and API (current state) or a negligible overhead.

> There are strictly NO internal access between assemblies.

About the link i was thinking a build hack to enable `InternalsVisibleTo` which would be not neat for core devs but very valuable for end users.

**Reply by colejohnson66** [reply]
> Even if they are in different assemblies it should be possible, worse case with an internal (copy on write pattern to avoid locks) Dictionary<Predicate, TryAddDelegate> the concurrent dictionary can register statically (when class is loaded) in the extension, this way it will behave as expected even if modules are not the same and this dictionary can be opened for concurrent assembly but not end users to enable to break it when a proper solution is found. So clearly there are multiple solution without even reflection and with low enough cost - dictionary option can even just be a dedicated static attribute if concurrent dictionary is the only case, overhead will be pretty much 0 and it will be functional for end users.
> 
> Nothing is never impossible in dev ;)

A static dictionary of type-&gt;delegate would itself need to be a `ConcurrentDictionary` to ensure safety in registration. But then you're back to square one because the current runtime library setup doesn't allow `System.Collections.dll` to use anything in `System.Collections.Concurrent.dll`.

Sure, it might be possible with `TypeForwarded[From|To]` attributes and some clever work, but that's a relatively large task with relatively little return. As shown, `TryAdd` is nothing more than a helper for `!HasKey > Add`. If you need concurrency checks, you can write your own extension:

```csharp
public static class MyCollectionExtensions
{
    public static bool SafeTryAdd<TKey, TValue>(this IDictionary<TKey, TValue> d, TKey key, TValue value)
    {
        if (d is ConcurrentDictionary<TKey, TValue> d2)
            return d2.TryAdd(key, value);

        return d.TryAdd(key, value); // fallback to runtime impl
   }
}
```

Combine that with the [banned API analyzer](https://github.com/dotnet/roslyn-analyzers/blob/main/src/Microsoft.CodeAnalysis.BannedApiAnalyzers/BannedApiAnalyzers.Help.md), and you can enforce this usage in your own codebase.

**Reply by rmannibucau** [reply]
> A static dictionary of type->delegate would itself need to be a ConcurrentDictionary to ensure safety in registration.

Not really since it is internal the usage can be guaranteed (can't be a Dictionary, ack on this but still be can an array or just dedicated handlers to impl depending the number of impl to support - if only concurrent one it is 1 and no need of associative array at all).

That said there are other way to impl that, maybe using an internal (+ visible to runtime assemblies) interface or alike.
Otherwise goal is to enable to couple both assemblies to make it right by default without forcing end user to redevelop the provided extensions.


> Sure, it might be possible with TypeForwarded[From|To] attributes and some clever work, but that's a relatively large task with relatively little return. As shown, TryAdd is nothing more than a helper for !HasKey > Add. If you need concurrency checks, you can write your own extension

Don't think ROI is low, why does TryAdd not use TryAdd (from an user standpoint it is literally that). Concurrent dictionaries are more and more common and this API is very misleading.
Indeed you can also bypass existing API and do your own but what's the point?

**Reply by huoyaoyuan** [reply]
> why does TryAdd not use TryAdd

A generic entry can never use specialized one without some sort of polymorphism. The real solution of the problem can only be polymorphism.

> Concurrent dictionaries are more and more common and this API is very misleading.

Using concurrent dictionary as `IDictionary` is not common. There are many more algorithms assuming single thread can break. That's why I was suggesting to catch all the casts. User should review and suppress cases which are determined to be safe.

> Don't think ROI is low

It will never be considered to break all the rules to benefit only a single method, which shouldn't be used in best pattern.

**Reply by rmannibucau** [reply]
> The real solution of the problem can only be polymorphism.

agree but you said it will not happen anytime soon so we need something else

> Using concurrent dictionary as IDictionary is not common. There are many more algorithms assuming single thread can break. That's why I was suggesting to catch all the casts. User should review and suppress cases which are determined to be safe.

Yes and not, when you assume you own all the code and all the methods you call, maybe, but as soon as you are using libs it depends more and shouldn't be assumed IMHO

> It will never be considered to break all the rules to benefit only a single method, which shouldn't be used in best pattern.

s/single/extension methods/

also if it shouldn't be used it shouldn't be there or be obsolete probably

once again c# has interfaces and inheritances but extensions methods don't support it - until there it is fair, but there is no way to identify it from a code perspective until you own all the code of all you bundle which is almost never the case or is not practicable (an analyzer analyzing all the transitive deps for ex)
so it needs a proper short term solution until extension polymorphism is there IMHO, it creates real prod issues almost impossible to detect at build time
